File: snobot_sim/cpp/lib/SnobotSim/MotorFactory/VexMotorFactory.cpp
/*
 * VexMotorFactory.cpp
 *
 *  Created on: Jun 21, 2017
 *      Author: preiniger
 */

#include "SnobotSim/MotorFactory/VexMotorFactory.h"
#include "SnobotSim/Logging/SnobotLogger.h"


const std::string VexMotorFactory::MOTOR_NAME_CIM            = "CIM";
const std::string VexMotorFactory::MOTOR_NAME_MINI_CIM       = "Mini CIM";
const std::string VexMotorFactory::MOTOR_NAME_BAG            = "Bag";
const std::string VexMotorFactory::MOTOR_NAME_775_PRO        = "775 Pro";
const std::string VexMotorFactory::MOTOR_NAME_AM_RS_775_125  = "Andymark RS 775-125";
const std::string VexMotorFactory::MOTOR_NAME_BB_RS_775      = "Banebots RS 775";
const std::string VexMotorFactory::MOTOR_NAME_AM_9015        = "Andymark 9015";
const std::string VexMotorFactory::MOTOR_NAME_BB_RS_550      = "Banebots RS 550";

VexMotorFactory::VexMotorFactory()
{

}

VexMotorFactory::~VexMotorFactory()
{

}

DcMotorModelConfig VexMotorFactory::MakeTransmission(
        const DcMotorModelConfig& aMotorConfig, int aNumMotors, double aReduction, double aEfficiency)
{
    DcMotorModelConfig::FactoryParams factoryParams;
    factoryParams.mMotorName = aMotorConfig.mFactoryParams.mMotorName;
    factoryParams.mNumMotors = aNumMotors;
    factoryParams.mGearReduction = aReduction;
    factoryParams.mTransmissionEfficiency = aEfficiency;

    DcMotorModelConfig output(
            factoryParams,
            aMotorConfig.NOMINAL_VOLTAGE,
            aMotorConfig.FREE_SPEED_RPM / aReduction,
            aMotorConfig.FREE_CURRENT  * aNumMotors,
            aMotorConfig.STALL_TORQUE  * aNumMotors,
            aMotorConfig.STALL_CURRENT * aNumMotors,
            aMotorConfig.mMotorInertia * aNumMotors * aReduction * aReduction,
            aMotorConfig.mHasBrake,
            aMotorConfig.mInverted
            );


    output.mKT *= aEfficiency * aNumMotors * aReduction;

    return output;
}

DcMotorModelConfig VexMotorFactory::CreateMotor(
        const std::string& aName)
{
    const double NOMINAL_VOLTAGE = 12;

    DcMotorModelConfig::FactoryParams factoryParams;
    factoryParams.mMotorName = aName;
    factoryParams.mNumMotors = 1;
    factoryParams.mGearReduction = 1;
    factoryParams.mTransmissionEfficiency = 1;

    if(aName == MOTOR_NAME_CIM)
    {
        const double FREE_SPEED_RPM = 5330;
        const double FREE_CURRENT = 2.7;
        const double STALL_TORQUE = 2.41;
        const double STALL_CURRENT = 131;

        return DcMotorModelConfig(factoryParams, NOMINAL_VOLTAGE, FREE_SPEED_RPM, FREE_CURRENT, STALL_TORQUE, STALL_CURRENT, 0);
    }
    else if(aName == MOTOR_NAME_MINI_CIM)
    {
        const double FREE_SPEED_RPM = 5840;
        const double FREE_CURRENT = 3;
        const double STALL_TORQUE = 1.41;
        const double STALL_CURRENT = 89;

        return DcMotorModelConfig(factoryParams, NOMINAL_VOLTAGE, FREE_SPEED_RPM, FREE_CURRENT, STALL_TORQUE, STALL_CURRENT, 0);
    }
    else if(aName == MOTOR_NAME_BAG)
    {
        const double FREE_SPEED_RPM = 13180;
        const double FREE_CURRENT = 1.8;
        const double STALL_TORQUE = .43;
        const double STALL_CURRENT = 53;

        return DcMotorModelConfig(factoryParams, NOMINAL_VOLTAGE, FREE_SPEED_RPM, FREE_CURRENT, STALL_TORQUE, STALL_CURRENT, 0);
    }
    else if(aName == MOTOR_NAME_775_PRO)
    {
        const double FREE_SPEED_RPM = 18730;
        const double FREE_CURRENT = .7;
        const double STALL_TORQUE = .71;
        const double STALL_CURRENT = 134;

        return DcMotorModelConfig(factoryParams, NOMINAL_VOLTAGE, FREE_SPEED_RPM, FREE_CURRENT, STALL_TORQUE, STALL_CURRENT, 0);
    }
    else if(aName == MOTOR_NAME_AM_RS_775_125)
    {
        const double FREE_SPEED_RPM = 5800;
        const double FREE_CURRENT = 1.6;
        const double STALL_TORQUE = .28;
        const double STALL_CURRENT = 18;

        return DcMotorModelConfig(factoryParams, NOMINAL_VOLTAGE, FREE_SPEED_RPM, FREE_CURRENT, STALL_TORQUE, STALL_CURRENT, 0);
    }
    else if(aName == MOTOR_NAME_BB_RS_775)
    {
        const double FREE_SPEED_RPM = 13050;
        const double FREE_CURRENT = 2.7;
        const double STALL_TORQUE = .72;
        const double STALL_CURRENT = 97;

        return DcMotorModelConfig(factoryParams, NOMINAL_VOLTAGE, FREE_SPEED_RPM, FREE_CURRENT, STALL_TORQUE, STALL_CURRENT, 0);
    }
    else if(aName == MOTOR_NAME_AM_9015)
    {
        const double FREE_SPEED_RPM = 14270;
        const double FREE_CURRENT = 3.7;
        const double STALL_TORQUE = .36;
        const double STALL_CURRENT = 71;

        return DcMotorModelConfig(factoryParams, NOMINAL_VOLTAGE, FREE_SPEED_RPM, FREE_CURRENT, STALL_TORQUE, STALL_CURRENT, 0);
    }
    else if(aName == MOTOR_NAME_BB_RS_550)
    {
        const double FREE_SPEED_RPM = 19000;
        const double FREE_CURRENT = .4;
        const double STALL_TORQUE = .38;
        const double STALL_CURRENT = 84;

        return DcMotorModelConfig(factoryParams, NOMINAL_VOLTAGE, FREE_SPEED_RPM, FREE_CURRENT, STALL_TORQUE, STALL_CURRENT, 0);
    }
    else
    {
        SNOBOT_LOG(SnobotLogging::ERROR, "Unknown motor sim name " << aName);
    }

    factoryParams.mMotorName = "";
    return DcMotorModelConfig(factoryParams, 0, 0, 0, 0, 0, 0);
}

File: snobot_sim_java/java/snobot_sim_java/com/snobot/simulator/motor_sim/motor_factory/VexMotorFactory.java
package com.snobot.simulator.motor_sim.motor_factory;

import java.util.HashMap;
import java.util.Map;

import com.snobot.simulator.motor_sim.DcMotorModelConfig;

public class VexMotorFactory
{
    private final static double NOMINAL_VOLTAGE = 12;

    protected static final Map<String, DcMotorModelConfig.MotorParams> sMOTOR_PARAMS;

    protected static final DcMotorModelConfig.MotorParams MOTOR_PARAMS_CIM = new DcMotorModelConfig.MotorParams(NOMINAL_VOLTAGE, 5330, 2.7, 2.41, 131);
    protected static final DcMotorModelConfig.MotorParams MOTOR_PARAMS_MINI_CIM = new DcMotorModelConfig.MotorParams(NOMINAL_VOLTAGE, 5840, 3, 1.41, 89);
    protected static final DcMotorModelConfig.MotorParams MOTOR_PARAMS_BAG = new DcMotorModelConfig.MotorParams(NOMINAL_VOLTAGE, 13180, 1.8, .43, 53);
    protected static final DcMotorModelConfig.MotorParams MOTOR_PARAMS_775PRO = new DcMotorModelConfig.MotorParams(NOMINAL_VOLTAGE, 18730, .7, .71, 134);
    protected static final DcMotorModelConfig.MotorParams MOTOR_PARAMS_AM_775_125 = new DcMotorModelConfig.MotorParams(NOMINAL_VOLTAGE, 5800, 1.6, .28, 18);
    protected static final DcMotorModelConfig.MotorParams MOTOR_PARAMS_BB_RS775 = new DcMotorModelConfig.MotorParams(NOMINAL_VOLTAGE, 13050, 2.7, .72, 97);
    protected static final DcMotorModelConfig.MotorParams MOTOR_PARAMS_AM_9015 = new DcMotorModelConfig.MotorParams(NOMINAL_VOLTAGE, 14270, 3.7, .36, 71);
    protected static final DcMotorModelConfig.MotorParams MOTOR_PARAMS_BB_RS550 = new DcMotorModelConfig.MotorParams(NOMINAL_VOLTAGE, 19000, .4, .38, 84);

    static
    {
        sMOTOR_PARAMS = new HashMap<>();
        sMOTOR_PARAMS.put("CIM", MOTOR_PARAMS_CIM);
        sMOTOR_PARAMS.put("Mini CIM", MOTOR_PARAMS_MINI_CIM);
        sMOTOR_PARAMS.put("Bag", MOTOR_PARAMS_BAG);
        sMOTOR_PARAMS.put("775 Pro", MOTOR_PARAMS_775PRO);
        sMOTOR_PARAMS.put("Andymark RS 775-125", MOTOR_PARAMS_AM_775_125);
        sMOTOR_PARAMS.put("Banebots RS 775", MOTOR_PARAMS_BB_RS775);
        sMOTOR_PARAMS.put("Andymark 9015", MOTOR_PARAMS_AM_9015);
        sMOTOR_PARAMS.put("Banebots RS 550", MOTOR_PARAMS_BB_RS550);
    }

    public static DcMotorModelConfig createMotor(String aMotorName)
    {
        DcMotorModelConfig.FactoryParams factoryParams = new DcMotorModelConfig.FactoryParams(aMotorName, 1, 1, 1);
        DcMotorModelConfig.MotorParams motorParams = sMOTOR_PARAMS.get(aMotorName);

        if (motorParams != null)
        {
            return new DcMotorModelConfig(factoryParams, motorParams, false, false);
        }
        return new DcMotorModelConfig(factoryParams, new DcMotorModelConfig.MotorParams(NOMINAL_VOLTAGE, 0, 0, 0, 0), false, false);
    }
}
